pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,423 @@
+	tv.logf(1, "registering remote trace %s", tv)
+}
+
+// withContext allows us to get the context from the object.
+type withContext interface {
+	Context() context.Context
+}
+
+// MaybeRegister is called on an entry that is about to be proposed. This will
+// begin logging all subsequent updates to this entry. It will return true if
+// the registration succeeded.
+func (r *RaftTracer) MaybeRegister(propCtx withContext, ent raftpb.Entry) bool {
+	// Only register if there is a trace in the context and it is set to verbose
+	// logging.
+	span := tracing.SpanFromContext(propCtx.Context())
+	if span == nil || span.RecordingType() != tracingpb.RecordingVerbose {
+		return false
+	}
+
+	// If the index is nil, then we can't trace this entry. This can happen if
+	// there is a leader/leaseholder spilt. We don't have an easy way to handle
+	// this today, so don't attempt to trace it.
+	if ent.Index == 0 {
+		log.VEventf(propCtx.Context(), 2, "skip registering raft proposal without index: %v", ent)
+		return false
+	}
+
+	// This must be the last conditional. If this returns true we must call
+	// storeEntry to not leak a registered permit.
+	if !r.reserveSpace() {
+		log.VEvent(propCtx.Context(), 2, "too many active raft traces, skipping")
+		return false
+	}
+
+	// Grab the trace and span id now because the propCtx.Context may be
+	// nil'ed later.
+	tv := r.storeEntry(
+		kvserverpb.TracedEntry{
+			Index:   kvpb.RaftIndex(ent.Index),
+			TraceID: span.TraceID(),
+			SpanID:  span.SpanID(),
+		},
+		propCtx,
+	)
+	tv.logf(1, "registering local trace %s", tv)
+	return true
+}
+
+// MaybeTrace logs the message if it is covered by a trace.
+func (r *RaftTracer) MaybeTrace(m raftpb.Message) []kvserverpb.TracedEntry {
+	// NB: This check is an optimization to handle the common case where there
+	// are no registered traces on the store.
+	if r.numRegisteredReplica.Load() == 0 {
+		return nil
+	}
+
+	switch m.Type {
+	case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+		return r.traceIfCovered(m)
+	case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp, raftpb.MsgStorageApplyResp:
+		r.traceIfPast(m)
+		return nil
+	}
+	return nil
+}
+
+// removeEntry remove this entry and decrements the registered counters at the
+// replica and store level.
+func (r *RaftTracer) removeEntry(index kvpb.RaftIndex) {
+	r.m.Delete(index)
+	r.numRegisteredStore.Add(-1)
+	r.numRegisteredReplica.Add(-1)
+}
+
+// Close will unregister all the currently active traces and prevent additional
+// traces from being added. It is safe to call multiple times, but should always
+// be called at least once when the replica is destroyed to prevent leaking
+// traces.
+func (r *RaftTracer) Close() {
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		r.removeEntry(index)
+		t.logf(2, "cleanup log index %d during FlushAll", index)
+		return true
+	})
+	// Prevent any new registrations.
+	r.numRegisteredReplica.Store(math.MaxInt64)
```

Comment:
This should go first. Otherwise, if additions are racing with deletions, we still can end up with a non-empty map at the end.

Reply:
I moved the to front and added a comment. Since we first set to max int and then remove entries, it will drop below max int if there were any registered, but I think that is OK. This is still not 100% bulletproof though. In `MaybeRegister` we can succeed in `reserveSpace` then we call `Close` and then we finally call `storeEntry`. We won't leak any permits, but the entry will still be in the local map. That is probably OK since we won't re-increment the counter, but this trace may get left dangling in the replica's RaftTracer. 

The good news is that now we shouldn't have anything in `numRegisteredStore`.

Reply:
Actually this might be somewhat problematic now. If we update this while someone is in the middle of `reserveSpace` we then later add 1 to it. This would mean that we went from +MAX -> negative. Future checks will always fail. i'm going to use a smaller number (max32) to prevent this case at least....